### DISCUSSION TITLE
XCode 12.x example?

### DISCUSSION THREAD

**Discussion by Planet30** [brampf/fitscore#8] [Q&A] [answered]
Do you happen to have an example for XCode 12.x with Swift 5.3 for iPadOS 14.x and/or macOS 11.x? I noticed, that the provided LinuxTest has a few error messages in XCode. I could dig in... but if somebody has done this already that would save me some time. Thank you!

**Comment by Planet30** [comment]
> Du you happen to have an example for XCode 12.x with Swift 5.3 for iPadOS 14.x and/or macOS 11.x? I noticed, that the provided LinuxTest has a few error messages in XCode. I could dig in... but if somebody has done this already that would save me some time. Thank you!

I will try out fitskit and see if it works for me ...

**Comment by brampf** [comment, marked as answer]
Hi again,

I'm afraid fitskit isn't as sophisticated as I intended it to be. But both fitscore and fitskit sources and tests do compile and run without any errors in my test environments which are
* macOS 11.1 (x86) + Swift 5.3.2
* Ubuntu 20.04.1 (x86) + Swift 5.3.1 (this is also the environment used for performing the checks here on GitHub)

Please note that if you run the tests in the iOS simulator the writer tests will complain that they cannot write the sample files due to limitations in the simulator. I'm intending to fix this some day.

I have both projects all set up and running in a larger Xcode project which compiles (and runs ;) on iPadOS 14.3. Unfortunately I cannot share this project as it contains code I cannot publish. If you struggle with setting up a project in Xcode using SPM, I recommend having a look at https://developer.apple.com/documentation/xcode/adding_package_dependencies_to_your_app .

**Reply by Planet30** [reply]
Thanks for the reply. I could make it work at last.... Can load and show an FITS image using SwiftUI.  The completion handler gave me some headaches in file.prime.cgimage(...). I had to get the cgimage out of that handler.

Sorry, that I bothered you with these trivial issues. Still learning Swift and SwiftUI.

**Reply by brampf** [reply]
No need to apologise. We've all been there.
The reason for the completion handler being that way is so that I can make use of accelerate to assemble the image. I know it's not smooth with SwiftUI as you have to write a piece of async code. But rendering the image may take a few seconds and therefore you definitely shouldn't do this in the primary run loop.

**Reply by Planet30** [reply]
Next challenge ... My FITS images are in the RGGB pixel format from an astrophotography camera (ASI, using a Sony color chip).

1) I like to convert the FITS with either CoreImage or with CVPixelBufferCreate... to a raw image in order to apply all the nice RawFilters from iOS/Swift to them.

I'm completely stuck. I can create a pixel buffer with the kCVPixelFormaType_14Bayer_RGGB, but the quickview in XCode only shows me a grayscale image. I am not able to convert it to a CIImage or CGImage showing colors, it is either nil or remains a grayscale image that is not de-bayered / demosaiced.

2) Looking at your RGB conversion of the FITS (FITSBytesTool), I noticed the RGB are consecutively ordered (3x layersize). This is not the case for me, where I have a Bayer matrix. Re-writing this function to separate the pixel colors, does result in a color image, but the colors are all wrong...

Any hint to solve either problem would be highly appreciated.

**Reply by brampf** [reply]
I created issue #10 for this
